segmentstore/server/src/test/java/io/pravega/segmentstore/server/attributes/ContainerAttributeIndexTests.java

```diff
@@ -0,0 +1,698 @@
+
+        AssertExtensions.assertThrows(
+                "put() worked after delete().",
+                () -> idx.put(UUID.randomUUID(), 0L, TIMEOUT),
+                ex -> ex instanceof StreamSegmentNotExistsException);
+
+        // Check index before sealing.
+        checkIndex(idx, Collections.emptyMap());
+    }
+
+    /**
+     * Tests the case when Snapshots fail due to a Storage failure. This should prevent whatever operation triggered it
+     * to completely fail and not record the data.
+     */
+    @Test
+    public void testStorageFailure() {
+        val attributeId = UUID.randomUUID();
+        val lastWrittenValue = new AtomicLong(0);
+        val config = AttributeIndexConfig.builder()
+                                         .with(AttributeIndexConfig.UPDATE_COUNT_THRESHOLD_SNAPSHOT, 4)
+                                         .build();
+        @Cleanup
+        val context = new TestContext(config);
+        populateSegments(context);
+        val idx = (SegmentAttributeIndex) context.index.forSegment(SEGMENT_ID, TIMEOUT).join();
+
+        // 1. When writing normally
+        context.storage.writeInterceptor = (streamSegmentName, offset, data, length, wrappedStorage) -> {
+            throw new IntentionalException();
+        };
+        AssertExtensions.assertThrows(
+                "put() worked with Storage failure.",
+                () -> idx.put(attributeId, 0L, TIMEOUT),
+                ex -> ex instanceof IntentionalException);
+        Assert.assertNull("A value was retrieved after a failed put().", idx.get(attributeId, TIMEOUT).join());
+
+        // 2. When generating a snapshot.
+        // Write as much as we can until we are about to create a snapshot (but don't do it yet).
+        context.storage.writeInterceptor = null;
+        while (!idx.shouldSnapshot()) {
+            idx.put(attributeId, lastWrittenValue.incrementAndGet(), TIMEOUT).join();
+        }
+
+        context.storage.writeInterceptor = (streamSegmentName, offset, data, length, wrappedStorage) -> {
+            throw new IntentionalException();
+        };
+        AssertExtensions.assertThrows(
+                "put()/snapshot worked with Storage failure.",
+                () -> idx.put(attributeId, lastWrittenValue.get() + 1, TIMEOUT),
+                ex -> ex instanceof IntentionalException);
+        Assert.assertEquals("An invalid value was retrieved after a failed put()/snapshot.",
+                lastWrittenValue.get(), (long) idx.get(attributeId, TIMEOUT).join());
+
+        AssertExtensions.assertThrows(
+                "seal() worked with Storage failure.",
+                () -> idx.seal(TIMEOUT),
+                ex -> ex instanceof IntentionalException);
+    }
+
+    /**
+     * Tests the case when Snapshots fail due to OperationLogFailure. The snapshot itself must have recorded in the file,
+     * but the Offset & Length attributes never updated on the main Segment.
```

Comment:
Better not to use ampersand as it might confuse the javadoc compiler. Use `and` directly.

Reply:
fixed